modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,596 @@
+import static org.apache.ignite.internal.util.distributed.DistributedProcess.DistributedProcessType.SNAPSHOT_CHECK_METAS;
+import static org.apache.ignite.internal.util.distributed.DistributedProcess.DistributedProcessType.SNAPSHOT_VALIDATE_PARTS;
+
+/** Distributed process of snapshot checking (with the partition hashes). */
+public class SnapshotCheckProcess {
+    /** */
+    private final IgniteLogger log;
+
+    /** */
+    private final GridKernalContext kctx;
+
+    /** Operation contexts by name. */
+    private final Map<String, SnapshotCheckContext> contexts = new ConcurrentHashMap<>();
+
+    /** Cluster-wide operation futures per snapshot called from current node. */
+    private final Map<UUID, GridFutureAdapter<SnapshotPartitionsVerifyTaskResult>> clusterOpFuts = new ConcurrentHashMap<>();
+
+    /** Check metas first phase subprocess. */
+    private final DistributedProcess<SnapshotCheckProcessRequest, SnapshotCheckResponse> phase1CheckMetas;
+
+    /** Partition hashes second phase subprocess.  */
+    private final DistributedProcess<SnapshotCheckProcessRequest, SnapshotCheckResponse> phase2PartsHashes;
+
+    /** */
+    public SnapshotCheckProcess(GridKernalContext kctx) {
+        this.kctx = kctx;
+
+        log = kctx.log(getClass());
+
+        phase1CheckMetas = new DistributedProcess<>(kctx, SNAPSHOT_CHECK_METAS, this::prepareAndCheckMetas,
+            this::reducePreparationAndMetasCheck);
+
+        phase2PartsHashes = new DistributedProcess<>(kctx, SNAPSHOT_VALIDATE_PARTS, this::validateParts,
+            this::reduceValidatePartsAndFinish);
+
+        kctx.event().addLocalEventListener(evt -> onNodeLeft(((DiscoveryEvent)evt).eventNode().id()), EVT_NODE_FAILED, EVT_NODE_LEFT);
+    }
+
+    /** Expected to run in a discovery-managed thread. */
+    private void onNodeLeft(UUID nodeId) {
+        Throwable err = new ClusterTopologyCheckedException("Snapshot validation stopped. A required node left the cluster " +
+            "[nodeId=" + nodeId + ']');
+
+        Iterator<Map.Entry<String, SnapshotCheckContext>> it = contexts.entrySet().iterator();
+
+        while (it.hasNext()) {
+            SnapshotCheckContext ctx = it.next().getValue();
+
+            if (!ctx.req.nodes().contains(nodeId))
+                continue;
+
+            if (ctx.fut != null)
+                ctx.fut.onDone(err);
+
+            it.remove();
+
+            GridFutureAdapter<?> fut = clusterOpFuts.get(ctx.req.reqId);
+
+            if (fut != null)
+                fut.onDone(err);
+        }
+    }
+
+    /** */
+    Map<String, SnapshotCheckContext> contexts() {
+        return Collections.unmodifiableMap(contexts);
+    }
+
+    /**
+     * Stops all the processes with the passed exception.
+     *
+     * @param err The interrupt reason.
+     */
+    void interrupt(Throwable err) {
+        contexts.forEach((snpName, ctx) -> {
+            if (ctx.fut != null)
+                ctx.fut.onDone(err);
```

Comment:
It is possible `interrupt` will not interrupt anything. You still have a period of time:
1. First phase don't start and futures are nulls.
2. First phase finished, second phase still not started and futures are nulls.

I again recommend you create a future that lives on every node during all process and phase futures listen it.

Reply:
This call in in the same thread and discovery managed. We either start or reduce any phase or process this interruption. Absence of the futures during the interuption doesn't look like problem. It means only that we have nothing to stop this time. Isn't it?

Reply:
`interrupt()` is a node stopping. SnapshotManager stops, its thread pool stops. Discovery should not work, process phases should start and be able to work because SnapshotManager stops. No problem is expected. Also, keeping single future requires resseting it. The same race. Canceling done future won't do anything. And the `reset()` would revive and restart it.